### DISCUSSION TITLE
Where has read_buf gone?

### DISCUSSION THREAD

**Discussion by kmod-midori** [tokio-rs/tokio#2970] [Q&A] [answered]
Based on #2735, I assume that they would be moved to `tokio-util`, which just got a new release hours ago. However, they are not there, which seems to leave me with no way to read into any kind of uninitialized buffer. `ReadBuf` looks great, but it seems that it is not meant to be directly used by end users (i.e. with no `async` API that reads into it)?

**Comment by Darksonn** [comment]
We intend to provide a version of it in tokio-util soon, but it did not make it for the 0.4.0 release of tokio-util. In the mean-time, you would need to go through the `ReadBuf` API manually.

**Comment by Darksonn** [comment, marked as answer]
Tokio 0.3.1 has now been released with `read_buf` and `write_buf`. You can find the changelog [here](https://github.com/tokio-rs/tokio/releases/tag/tokio-0.3.1).
